invokeai/app/services/shared/sqlite/sqlite_migrator.py

```diff
@@ -0,0 +1,307 @@
+import shutil
+import sqlite3
+import threading
+from datetime import datetime
+from logging import Logger
+from pathlib import Path
+from typing import Callable, Optional, TypeAlias
+
+from pydantic import BaseModel, Field, model_validator
+
+MigrateCallback: TypeAlias = Callable[[sqlite3.Cursor], None]
+
+
+class MigrationError(RuntimeError):
+    """Raised when a migration fails."""
+
+
+class MigrationVersionError(ValueError):
+    """Raised when a migration version is invalid."""
+
+
+class Migration(BaseModel):
+    """
+    Represents a migration for a SQLite database.
+
+    Migration callbacks will be provided an open cursor to the database. They should not commit their
+    transaction; this is handled by the migrator.
+
+    Pre- and post-migration callback may be registered with :meth:`register_pre_callback` or
+    :meth:`register_post_callback`.
+
+    If a migration has additional dependencies, it is recommended to use functools.partial to provide
+    the dependencies and register the partial as the migration callback.
+    """
+
+    from_version: int = Field(ge=0, strict=True, description="The database version on which this migration may be run")
+    to_version: int = Field(ge=1, strict=True, description="The database version that results from this migration")
```

Comment:
Migrations are now aware of what version they can migrate from and to. This makes it much, much harder to bork.